## 1. Load the data

In [1]:
# imports:
import pandas as pd
import matplotlib.pyplot as plt

INPUT_CSV = '/home/sramkova/diploma_thesis_data/neighbourhood_both_days.csv'


raw_data = pd.read_csv(INPUT_CSV)

# check that all columns are correctly loaded
raw_data.head()

/home/sramkova/.local/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3165: DtypeWarning: Columns (22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


originated_ip    uid      connection.uid connection.conn_state  \
0     10.2.4.44  0x10e  C91B543mPc9uqyAmZg                    SF   
1     10.2.4.44  0x37a  CJDeQn1RA4RpXDJi76                    SF   
2     10.2.4.44  0x3c0   CPm0gG6Cmt4bLshDh                    SF   
3     10.2.4.44  0x7db   CFPUgMgxMkSXwpF0l                    S1   
4     10.2.4.44  0x80c  C93t9a4VGVuS6lYHD8                    SF   

   connection.duration  connection.orig_bytes  connection.orig_ip_bytes  \
0             0.523709                    603                      1727   
1             0.010425                    116                       384   
2             0.012546                    116                       384   
3             0.014672                    114                       277   
4             0.009000                    116                       384   

   connection.orig_p  connection.orig_pkts connection.proto  ...  \
0              52240                    21              tcp  ...   
1              51558                     5              tcp  ...   
2              51822                     5              tcp  ...   
3              52580                     3              tcp  ...   
4              52498                     5              tcp  ...   

   resp_resp_similar_http_status_code_count  \
0                                         0   
1                                         0   
2                                         0   
3                                         0   
4                                         0   

   resp_resp_similar_http_user_agent_count  \
0                                        0   
1                                        0   
2                                        0   
3                                        0   
4                                        0   

   resp_resp_similar_ssl_version_count  resp_resp_similar_ssl_cipher_count  \
0                                    0                                   0   
1                                    0                                   0   
2                                    0                                   0   
3                                    0                                   0   
4                                    0                                   0   

  resp_resp_similar_ssl_curve_count  \
0                                 0   
1                                 0   
2                                 0   
3                                 0   
4                                 0   

  resp_resp_similar_ssl_validation_status_count  \
0                                             0   
1                                             0   
2                                             0   
3                                             0   
4                                             0   

  resp_resp_similar_files_source_count  resp_resp_similar_file_md5_count  \
0                                    0                                 0   
1                                    0                                 0   
2                                    0                                 0   
3                                    0                                 0   
4                                    0                                 0   

   attacker_label  attack_label  
0              No        Normal  
1              No        Normal  
2              No        Normal  
3              No        Normal  
4              No        Normal  

[5 rows x 219 columns]

In [2]:
pd.set_option('display.max_rows', None)
raw_data.dtypes

originated_ip                                     object
uid                                               object
connection.uid                                    object
connection.conn_state                             object
connection.duration                              float64
connection.orig_bytes                              int64
connection.orig_ip_bytes                           int64
connection.orig_p                                  int64
connection.orig_pkts                               int64
connection.proto                                  object
connection.resp_bytes                              int64
connection.resp_ip_bytes                           int64
connection.resp_p                                  int64
connection.resp_pkts                               int64
connection.service                                object
connection.ts                                     object
responded_ip                                      object
dns_count                      

In [3]:
pd.set_option('display.max_rows', 10)

In [4]:
raw_data = raw_data[['originated_ip', 
                     'responded_ip', 
                     'connection.uid', 
                     'connection.ts', 
                     'connection.resp_pkts', 
                     'connection.resp_ip_bytes', 
                     'connection.orig_p', 
                     'connection.duration',
                     'connection.resp_bytes',
                     'connection.orig_pkts',
                     'connection.proto',
                     'connection.service', 
                     'connection.orig_bytes',
                     'connection.conn_state',
                     'connection.resp_p',
                     'connection.orig_ip_bytes',
                     'dns_count',
                     'ssh_count', 
                     'http_count',
                     'ssl_count',
                     'files_count',
                    ]]
raw_data.head()

originated_ip responded_ip      connection.uid                connection.ts  \
0     10.2.4.44   4.122.55.2  C91B543mPc9uqyAmZg  2019-03-19T10:10:49.194278Z   
1     10.2.4.44   4.122.55.2  CJDeQn1RA4RpXDJi76  2019-03-19T10:58:49.150898Z   
2     10.2.4.44   4.122.55.2   CPm0gG6Cmt4bLshDh  2019-03-19T11:23:34.113264Z   
3     10.2.4.44   4.122.55.2   CFPUgMgxMkSXwpF0l  2019-03-19T12:35:34.077791Z   
4     10.2.4.44   4.122.55.2  C93t9a4VGVuS6lYHD8  2019-03-19T12:28:49.134506Z   

   connection.resp_pkts  connection.resp_ip_bytes  connection.orig_p  \
0                    20                     34507              52240   
1                     6                       851              51558   
2                     5                       799              51822   
3                     3                       164              52580   
4                     5                       799              52498   

   connection.duration  connection.resp_bytes  connection.orig_pkts  ...  \
0             0.523709                  32091                    21  ...   
1             0.010425                    531                     5  ...   
2             0.012546                    531                     5  ...   
3             0.014672                    523                     3  ...   
4             0.009000                    531                     5  ...   

  connection.service connection.orig_bytes  connection.conn_state  \
0                ssl                   603                     SF   
1               http                   116                     SF   
2               http                   116                     SF   
3               http                   114                     S1   
4               http                   116                     SF   

  connection.resp_p  connection.orig_ip_bytes  dns_count  ssh_count  \
0               443                      1727          0          0   
1                80                       384          0          0   
2                80                       384          0          0   
3                80                       277          0          0   
4                80                       384          0          0   

   http_count  ssl_count  files_count  
0           0          1            1  
1           1          0            1  
2           1          0            1  
3           1          0            0  
4           1          0            1  

[5 rows x 21 columns]

# 2. Get to know the data

### 2.0 Basic statistics

In [5]:
# types and num of non-null values in columns
raw_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 472345 entries, 0 to 472344
Data columns (total 21 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   originated_ip             472345 non-null  object 
 1   responded_ip              472345 non-null  object 
 2   connection.uid            472345 non-null  object 
 3   connection.ts             472345 non-null  object 
 4   connection.resp_pkts      472345 non-null  int64  
 5   connection.resp_ip_bytes  472345 non-null  int64  
 6   connection.orig_p         472345 non-null  int64  
 7   connection.duration       472345 non-null  float64
 8   connection.resp_bytes     472345 non-null  int64  
 9   connection.orig_pkts      472345 non-null  int64  
 10  connection.proto          472345 non-null  object 
 11  connection.service        472345 non-null  object 
 12  connection.orig_bytes     472345 non-null  int64  
 13  connection.conn_state     472345 non-null  o

In [6]:
# simple statistics of numerical attributes: 
raw_data.describe()

connection.resp_pkts  connection.resp_ip_bytes  connection.orig_p  \
count         472345.000000              4.723450e+05      472345.000000   
mean              16.523109              4.141463e+04       41580.853982   
std              496.531368              1.646461e+06       19028.068474   
min                0.000000              0.000000e+00           0.000000   
25%                0.000000              0.000000e+00       35165.000000   
50%                1.000000              8.400000e+01       48136.000000   
75%                4.000000              7.000000e+02       55619.000000   
max           101731.000000              2.506347e+08       65534.000000   

       connection.duration  connection.resp_bytes  connection.orig_pkts  \
count        472345.000000           4.723450e+05         472345.000000   
mean             12.989186           5.257307e+04             23.539294   
std             169.497371           2.166400e+06            802.812056   
min               0.000000           0.000000e+00              0.000000   
25%               0.000000           0.000000e+00              1.000000   
50%               0.006710           4.800000e+01              1.000000   
75%               0.308995           5.310000e+02              6.000000   
max           26718.986511           3.102849e+08         176170.000000   

       connection.orig_bytes  connection.resp_p  connection.orig_ip_bytes  \
count           4.723450e+05      472345.000000              4.723450e+05   
mean            4.430956e+04        2795.068213              2.272216e+03   
std             7.887787e+06        8330.279635              1.181989e+05   
min             0.000000e+00           0.000000              0.000000e+00   
25%             0.000000e+00          53.000000              4.400000e+01   
50%             4.200000e+01         443.000000              7.600000e+01   
75%             1.950000e+02        1080.000000              5.260000e+02   
max             1.981718e+09       65389.000000              4.545502e+07   

           dns_count      ssh_count     http_count      ssl_count  \
count  472345.000000  472345.000000  472345.000000  472345.000000   
mean        0.192874       0.039530       0.064434       0.191614   
std         0.394555       0.194853       0.245524       0.393571   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
max         1.000000       1.000000       1.000000       1.000000   

         files_count  
count  472345.000000  
mean        0.343061  
std         0.911913  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max       128.000000

In [7]:
# source: https://www.geeksforgeeks.org/python-pandas-dataframe-describe-method/

# remove null values to avoid errors  
nonnull_data = raw_data.dropna()  
  
# percentile list 
perc =[.20, .40, .60, .80] 
  
# list of dtypes to include 
include =['object', 'float', 'int'] 
  
# calling describe method 
nonnull_data.describe(percentiles = perc, include = include)

originated_ip responded_ip      connection.uid  \
count          472345       472345              472345   
unique            275         2678              472345   
top     212.96.119.83   4.122.55.3  CCsVpU14MAOGVsSYy2   
freq            28331        65859                   1   
mean              NaN          NaN                 NaN   
...               ...          ...                 ...   
40%               NaN          NaN                 NaN   
50%               NaN          NaN                 NaN   
60%               NaN          NaN                 NaN   
80%               NaN          NaN                 NaN   
max               NaN          NaN                 NaN   

                      connection.ts  connection.resp_pkts  \
count                        472345         472345.000000   
unique                       472345                   NaN   
top     2019-03-19T15:41:57.027409Z                   NaN   
freq                              1                   NaN   
mean                            NaN             16.523109   
...                             ...                   ...   
40%                             NaN              1.000000   
50%                             NaN              1.000000   
60%                             NaN              1.000000   
80%                             NaN              6.000000   
max                             NaN         101731.000000   

        connection.resp_ip_bytes  connection.orig_p  connection.duration  \
count               4.723450e+05      472345.000000        472345.000000   
unique                       NaN                NaN                  NaN   
top                          NaN                NaN                  NaN   
freq                         NaN                NaN                  NaN   
mean                4.141463e+04       41580.853982            12.989186   
...                          ...                ...                  ...   
40%                 7.300000e+01       42500.000000             0.003680   
50%                 8.400000e+01       48136.000000             0.006710   
60%                 1.640000e+02       50944.000000             0.019900   
80%                 2.191200e+03       57028.000000             0.661689   
max                 2.506347e+08       65534.000000         26718.986511   

        connection.resp_bytes  connection.orig_pkts  ... connection.service  \
count            4.723450e+05         472345.000000  ...             472345   
unique                    NaN                   NaN  ...                 14   
top                       NaN                   NaN  ...               none   
freq                      NaN                   NaN  ...             223400   
mean             5.257307e+04             23.539294  ...                NaN   
...                       ...                   ...  ...                ...   
40%              0.000000e+00              1.000000  ...                NaN   
50%              4.800000e+01              1.000000  ...                NaN   
60%              8.400000e+01              1.000000  ...                NaN   
80%              1.548000e+03              9.000000  ...                NaN   
max              3.102849e+08         176170.000000  ...                NaN   

       connection.orig_bytes  connection.conn_state connection.resp_p  \
count           4.723450e+05                 472345     472345.000000   
unique                   NaN                     13               NaN   
top                      NaN                     SF               NaN   
freq                     NaN                 191580               NaN   
mean            4.430956e+04                    NaN       2795.068213   
...                      ...                    ...               ...   
40%             2.200000e+01                    NaN         80.000000   
50%             4.200000e+01                    NaN        443.000000   
60%             5.000000e+01                    NaN        

### 2.1 Show distinct counts

In [8]:
distinct_counter = raw_data.apply(lambda x: len(x.unique()))
distinct_counter

originated_ip              275
responded_ip              2678
connection.uid          472345
connection.ts           472345
connection.resp_pkts      1245
                         ...  
dns_count                    2
ssh_count                    2
http_count                   2
ssl_count                    2
files_count                 44
Length: 21, dtype: int64

### 2.2 Visualize frequencies

#### Get Frequency Counts of Categorical Columns: 

- connection.proto
- connection.service
- connection.conn_state

(https://cmdlinetips.com/2018/02/how-to-get-frequency-counts-of-a-column-in-pandas-dataframe/)

In [9]:
pd.reset_option('display.max_rows')

In [10]:
raw_data['connection.proto'].value_counts()

tcp     319216
udp     125207
icmp     27922
Name: connection.proto, dtype: int64

In [11]:
raw_data['connection.service'].value_counts(dropna=False) # display null values as well

none        223400
dns          93172
ssl          90425
http         29097
ntp          28421
ssh           7344
dhcp           372
krb_tcp         42
dce_rpc         38
smtp            14
imap            11
ssl,imap         5
socks            3
pop3             1
Name: connection.service, dtype: int64

In [12]:
# plt.rcParams["figure.figsize"] = (20,3)
# fig, ax = plt.subplots()
# raw_data['connection.service'].value_counts().plot(ax=ax, kind='bar')
# graph visualization skipped (useless here)

In [13]:
raw_data['connection.conn_state'].value_counts()

SF        191580
S0        171351
RSTO       45493
OTH        31726
RSTR       20631
S1          3031
SHR         2735
S3          2626
S2          1873
SH           437
RSTRH        412
REJ          315
RSTOS0       135
Name: connection.conn_state, dtype: int64

#### Get Frequency Counts of Non-categorical Columns: 

- originated_ip
- responded_ip
- connection.duration
- connection.orig_pkts
- connection.orig_ip_bytes
- connection.resp_p
- connection.orig_bytes
- connection.resp_bytes
- connection.resp_ip_bytes
- connection.orig_p
- connection.resp_pkts

In [14]:
raw_data['originated_ip'].value_counts()

212.96.119.83     28331
80.93.182.183     27736
4.122.55.221      26417
5.23.147.206      24031
111.66.58.19      21768
                  ...  
77.51.92.44           1
74.125.133.188        1
188.114.116.1         1
9.66.2.1              1
4.122.55.113          1
Name: originated_ip, Length: 275, dtype: int64

In [15]:
# or print relative frequencies of unique values:
# (sometimes percentage is a better criterion then the count)

raw_data['originated_ip'].value_counts(normalize=True)

212.96.119.83     0.059979
80.93.182.183     0.058720
4.122.55.221      0.055927
5.23.147.206      0.050876
111.66.58.19      0.046085
                    ...   
77.51.92.44       0.000002
74.125.133.188    0.000002
188.114.116.1     0.000002
9.66.2.1          0.000002
4.122.55.113      0.000002
Name: originated_ip, Length: 275, dtype: float64

In [16]:
raw_data['responded_ip'].value_counts()

4.122.55.3         65859
4.122.55.2         45695
9.66.22.14         27322
9.66.22.12         24571
9.66.22.13         22238
                   ...  
173.255.254.232        1
13.32.28.12            1
81.15.133.186          1
18.195.73.96           1
104.96.90.200          1
Name: responded_ip, Length: 2678, dtype: int64

In [17]:
raw_data['connection.duration'].value_counts()

0.000000    160583
0.000015        57
0.000016        52
0.000014        42
0.003855        38
             ...  
3.218763         1
0.058045         1
0.334438         1
0.033412         1
4.885891         1
Name: connection.duration, Length: 187754, dtype: int64

In [18]:
pd.set_option('display.max_rows', None)
raw_data['connection.duration'].value_counts().nlargest(100)

0.000000    160583
0.000015        57
0.000016        52
0.000014        42
0.003855        38
0.004077        35
0.004240        35
0.004001        34
0.003817        34
0.004270        34
0.003716        34
0.004076        33
0.003906        33
0.003975        33
0.004353        32
0.003729        32
0.003851        32
0.004360        32
0.004004        32
0.004107        32
0.003814        32
0.000013        31
0.000005        31
0.004047        31
0.003921        31
0.004132        31
0.003981        31
0.003696        31
0.003937        31
0.003880        31
0.004088        31
0.004177        31
0.004089        30
0.004096        30
0.003895        30
0.004294        30
0.003877        30
0.003986        30
0.003945        30
0.003991        30
0.003983        29
0.004291        29
0.004242        29
0.004080        29
0.003701        29
0.003967        29
0.004263        29
0.004173        29
0.003640        29
0.004163        29
0.003567        29
0.004105        29
0.003944    

In [19]:
# raw_data['connection.orig_pkts'].value_counts()
raw_data['connection.orig_pkts'].value_counts().nlargest(40)

1     292723
3      18769
5      17704
2      15780
6      11005
10     10470
9       6699
11      6213
12      5963
4       5697
13      5401
17      4636
8       4580
14      4290
7       3478
16      3410
15      3109
0       2896
18      2891
19      2850
20      2112
21      1902
22      1670
23      1575
24      1530
25      1395
26      1174
27      1018
28       993
39       935
38       926
43       869
46       846
44       829
36       826
45       816
37       813
40       804
29       802
47       797
Name: connection.orig_pkts, dtype: int64

In [20]:
# raw_data['connection.orig_ip_bytes'].value_counts()
# raw_data['connection.orig_ip_bytes'].value_counts().nlargest(40)

In [21]:
# raw_data['connection.resp_p'].value_counts()
pd.set_option('display.max_rows', 30)
raw_data['connection.resp_p'].value_counts()

53       95339
443      84623
80       47815
123      28421
0        26116
         ...  
53136        1
30392        1
49521        1
57143        1
33501        1
Name: connection.resp_p, Length: 1326, dtype: int64

In [22]:
pd.set_option('display.max_rows', None)
raw_data['connection.resp_p'].value_counts().nlargest(40)

53       95339
443      84623
80       47815
123      28421
0        26116
22       19610
995      12238
4443      8387
5678      2392
993       2260
587       2245
143       2089
25        2048
110       2047
3         1291
8          575
8080       404
67         372
500        245
1900       231
49155      215
49152      215
49154      215
49157      214
49153      213
5355       212
3389       184
49158      181
1          177
49159      165
445        152
23         152
21         152
1723       151
9002       150
88         149
135        149
464        148
4380       148
9001       146
Name: connection.resp_p, dtype: int64

Categories will be:

- 21
- 22
- 53
- 80
- 123
- 443
- 3389
- 0 - 1023 *other well known ports*
- 1024 - 49151 *other registered*
- 49152 - 65535 *other dynamic*

Or two types: 

1.
- 21
- 22
- 53
- 80
- 123
- 443
- 3389
- other

2.
- 0 - 1023 *other well known ports*
- 1024 - 49151 *other registered*
- 49152 - 65535 *other dynamic*


In [23]:
# raw_data['connection.orig_bytes'].value_counts()
raw_data['connection.orig_bytes'].value_counts().nlargest(40)

0       186321
48       27847
56       25943
28       22780
22       10671
45        8020
206       7238
195       6122
1494      5570
60        5533
63        5204
387       4648
600       4634
381       4621
603       4604
378       4176
42        4126
397       4073
881       2923
817       2863
47        2768
116       2355
113       2288
52        2255
49        2236
50        2134
54        1932
40        1904
55        1867
53        1845
44        1510
39        1500
517       1491
46        1485
865       1458
43        1388
95        1379
41        1190
58        1174
12        1113
Name: connection.orig_bytes, dtype: int64

In [24]:
# raw_data['connection.resp_bytes'].value_counts()
raw_data['connection.resp_bytes'].value_counts().nlargest(40)

0         207456
48         25825
56         25001
77         14407
342         7902
2448        7238
1548        7073
2149        6318
45          5662
568         4692
100958      4607
559         4546
100774      4440
100884      4072
32291       3194
32091       3160
32188       2937
5040        2669
531         2365
522         2319
7803        1738
91          1147
12          1108
300         1106
396         1096
393         1052
375         1024
490          977
158          944
4436         933
463          926
168          829
87           825
1733         786
505          780
391          726
382          690
437          637
32150        634
31950        633
Name: connection.resp_bytes, dtype: int64

In [25]:
# raw_data['connection.resp_ip_bytes'].value_counts()
raw_data['connection.resp_ip_bytes'].value_counts().nlargest(40)

0       175101
76       25834
84       25042
105      14817
164      12557
44        8989
370       8155
73        5803
2664      4706
152       4454
3509      2466
700       2444
691       2334
799       2192
790       2120
1308      1956
1290      1939
3457      1691
2716      1623
112       1526
40        1453
1972      1187
119       1186
424       1128
421       1079
403       1075
2024      1047
186        968
491        960
488        936
115        853
1920       846
419        749
410        731
465        687
92         680
617        680
138        640
409        638
2076       600
Name: connection.resp_ip_bytes, dtype: int64

In [26]:
pd.set_option('display.max_rows', 30)
raw_data['connection.orig_p'].value_counts()

123      28244
8        25927
35164     3006
35165     3004
45311     3000
         ...  
25247        1
6699         1
37595        1
31271        1
31075        1
Name: connection.orig_p, Length: 49492, dtype: int64

In [27]:
# raw_data['connection.orig_p'].value_counts().nlargest(40)

In [28]:
# raw_data['connection.resp_pkts'].value_counts()
pd.set_option('display.max_rows', 30)
raw_data['connection.resp_pkts'].value_counts().nlargest(40)

0     175101
1     147585
3      23044
4      13154
5      11382
       ...  
35       552
36       492
37       461
38       370
39       274
Name: connection.resp_pkts, Length: 40, dtype: int64

In [29]:
raw_data['dns_count'].value_counts()

0    381242
1     91103
Name: dns_count, dtype: int64

In [30]:
raw_data['ssh_count'].value_counts()

0    453673
1     18672
Name: ssh_count, dtype: int64

In [31]:
raw_data['http_count'].value_counts()

0    441910
1     30435
Name: http_count, dtype: int64

In [32]:
raw_data['ssl_count'].value_counts()

0    381837
1     90508
Name: ssl_count, dtype: int64

In [33]:
raw_data['files_count'].value_counts()

0      370110
1       66284
2       18676
3       14823
4        1804
        ...  
29          1
28          1
24          1
19          1
128         1
Name: files_count, Length: 44, dtype: int64